esp/services/ws_dfu/ws_dfuXRefService.hpp

```diff
@@ -26,116 +26,135 @@
+            return;
+
+        xRefNode->setCluster(cluster);
+        writeNodeQueue(xRefNode);
+        m_sem.signal();
+    }
+
     virtual int run()
     {
         Link();
-        while(m_bRun)
+        while (!stopThread)
         {
             m_sem.wait();
-            if (m_pNodeQueue.ordinality() != 0)
-                RunXRef();
+            runXRef();
         }
         Release();
         return 0;
     }
 
-    void RunXRef()
+    void runXRef()
     {
-
         //catch all exceptions so we can signal for the new build to start
-        try{
-            SetRunningStatus(true);
-            Owned<IXRefNode> xRefNode  = (IXRefNode*)m_pNodeQueue.dequeue();
-            _CurrentClusterName.clear();
-            xRefNode->getCluster(_CurrentClusterName);
-            if (xRefNode->useSasha()) // if sasha processing just set submitted
-                xRefNode->setStatus("Submitted");
-            else {
-                Owned<IPropertyTree> tree = runXRefCluster(_CurrentClusterName.str(), xRefNode);
-                DBGLOG("finished run");
+        try
+        {
+            while (true)
+            {
+                Owned<IXRefNode> xRefNode  = readNodeQueue();
+                if (!xRefNode)
+                    break;
+
+                if (xRefNode->useSasha()) // if sasha processing just set submitted
+                    xRefNode->setStatus("Submitted");
+                else
+                {
+                    setRunningStatus(true);
+                    xRefNode->getCluster(currentClusterName);
+                    Owned<IPropertyTree> tree = runXRefCluster(currentClusterName.str(), xRefNode);
+                    DBGLOG("finished run XRef for %s", currentClusterName.str());
+                    currentClusterName.clear();
+                    setRunningStatus(false);
+                }
             }
-            SetRunningStatus(false);
         }
         catch(IException* e)
         {
             StringBuffer errorStr;
             e->errorMessage(errorStr);
-            ERRLOG("Exception thrown while running XREF: %s",errorStr.str());
+            ERRLOG("Exception thrown while running XREF: %s", errorStr.str());
             e->Release();
         }
         catch(...)
         {
             ERRLOG("Unknown Exception thrown from XREF");
         }
-        //Signal that we are ready to process another job if there is one....
-        m_sem.signal();
-
     }
 
-    virtual bool IsRunning()
+    virtual bool isRunning()
     {
-        CriticalSection(_boolMutex);
-        return bRunning;
+        CriticalBlock b(critRunningStatus);
+        return xRefRunning;
     }
-    virtual bool IsQueued(const char* Queue)
+    virtual bool isQueued(const char* clusterName)
     {
-        if(Queue == 0)
+        if (isEmptyString(clusterName))
             return false;
-        ForEachItemIn(x,m_pNodeQueue)
+
+        if (!currentClusterName.isEmpty() && streq(currentClusterName, clusterName))
```

Comment:
@wangkx is this called from a different thread then runXRef? Should currentClusterName be protected?

Reply:
Good point. But, protecting currentClusterName may not fix the problem. Let me think about it.